ros_ws/src/autonomous/src/wall_following.cpp

```diff
@@ -0,0 +1,336 @@
+#include <iostream>
+#include <array>
+#include <ros/ros.h>
+#include <ros/console.h>
+#include "sensor_msgs/LaserScan.h"
+#include "std_msgs/Float64.h"
+#include "drive_msgs/drive_param.h"
+#include "drive_msgs/pid_input.h"
+#include <math.h>
+
+
+float g_right_error = 0;
+float g_right_prev_error = 0;
+float g_right_corrected_angle = 0;
+
+float g_left_error = 0;
+float g_left_prev_error = 0;
+float g_left_corrected_angle = 0;
+
+class WallFollowing
+{
+	public:
+	WallFollowing();
+
+	private:
+ 	ros::NodeHandle nh_;
+
+	ros::Subscriber lidar_subscriber;	
+ 	void lidarCallback(const sensor_msgs::LaserScan::ConstPtr& lidar);
+
+	ros::Publisher pid_publisher;
+
+	
+	// void adjustSpeed(double speed);
+};
+
+WallFollowing::WallFollowing()
+{
+	lidar_subscriber = nh_.subscribe<sensor_msgs::LaserScan>("/racer/laser/scan", 1, &WallFollowing::lidarCallback, this);
+	pid_publisher = nh_.advertise< drive_msgs::pid_input >("/pid_input", 1);
+}
+
+/**
+ * @brief Using the lidar scans and a given angle theta
+ * this method returns the range at the given theta after doing boundary checks.
+ * 
+ * @param lidar 
+ * @param theta 
+ * @return float 
+ */
+float getRange(const sensor_msgs::LaserScan::ConstPtr& lidar, float theta)
+{	
+	// ranges size is 720, angle range is 270°
+	// so one step equals an angle of 0,375°
+	// to access range at angle theta, you have to divide the given theta by the step size
+	float range = lidar->ranges[theta/0.375];
+
+	// check for absurd or nan values
+	if(range > 4 || std::isinf(range))
+	{
+		range = 4;
+	}
+	return range;
+}
+
+/**
+ * @brief Checks if there is an unavoidable wall in front of the car using the lidar scans
+ * and returns a signal (0) to inhibit driving if it is the case.
+ * 
+ * @param lidar 
+ * @return float 
+ */
+float emergencyStop(const sensor_msgs::LaserScan::ConstPtr& lidar)
```

Comment:
Why return float? it seems like you only return 1 or 0 and only check whether or not the return value was not 0, might as well switch to `bool`.